### PR TITLE
resolve-lock: resolve legacy lock if safepoint < lastSafepoint and skip runGCJob()

### PR DESCRIPTION
**Pull request by joccau**
Signed-off-by: joccau <zak.zhao@pingcap.com>

### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tidb/issues/35428

Problem Summary:
If we set a large value to `gcLifeTime`, the `runGCJob()` will be skipped until the `safepoint` exceeds `gcLastRunTime`, 
Thus, the expired lock will not be resolved immediately.

### What is changed and how it works?
If the `runGcJob()` is skipped, it still needs to resolve locks with the expired ttl.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No code

### Manual test
We meet a cluster, in which the `gcLifeTime` is set 24h, and the locks exists all the time even these lock has expired.
This locks keep the checkpoint advancing in `log-backup`.
After patch the tidb-server with this PR, the locks has been resolved and the checkpoint advancing became normal.
origin_img_v2_305ccf96-42c2-40a7-9bc0-6c8b51f89a6g

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: store/gcworker/gc_worker.go [hunk 1]
```diff
@@ -309,13 +309,32 @@ func (w *GCWorker) leaderTick(ctx context.Context) error {
...
+
 	ok, safePoint, err := w.prepare(ctx)
-	if err != nil || !ok {
-		if err != nil {
-			metrics.GCJobFailureCounter.WithLabelValues("prepare").Inc()
-		}
+	if err != nil {
+		metrics.GCJobFailureCounter.WithLabelValues("prepare").Inc()
 		return errors.Trace(err)
+	} else if !ok {
```
Comment (3pointer) [inline store/gcworker/gc_worker.go:324]:
```suggestion
	} else if !ok && w.logBackupEnabled {
```

Comment (joccau) [reply store/gcworker/gc_worker.go:324]: > 

Whether w.logBackupEnable is true or false, It need return at line:334

**Review by MyonKeminta — DISMISSED** [review]

File: store/gcworker/gc_worker.go (hunk 1 above)
Comment (3pointer) [reply store/gcworker/gc_worker.go:324]: OK

File: store/gcworker/gc_worker.go [hunk 2]
```diff
@@ -309,12 +309,29 @@ func (w *GCWorker) leaderTick(ctx context.Context) error {
...
+		if w.logBackupEnabled {
+			tryResolveLocksTS, err := w.getTryResolveLocksTS()
+			if err != nil {
+				return errors.Trace(err)
+			}
+			if err = w.legacyResolveLocks(ctx, 0, tryResolveLocksTS, concurrency); err != nil {
```
Comment (3pointer) [inline store/gcworker/gc_worker.go:330]: you can add some comments about why we need these steps and why we can pass 0 here

**Review by sticnarf — APPROVED** [review]
LGTM

**Review by MyonKeminta — APPROVED** [review]

**Comment by wuhuizuo** [comment]
merged CI run failed

https://ci.pingcap.net/blue/organizations/jenkins/tidb_ghpr_coverage/detail/tidb_ghpr_coverage/3699/pipeline/71

@joccau

**Pull request merged**